Implement Serial Title API search

SolMe opened the pull request:
This PR adds an implementation of the first GET method of [Serial Title API](https://dev.elsevier.com/documentation/SerialTitleAPI.wadl#d1e534) of Scopus via `SerialSearch` class. `SerialSearch` allows to search against the serial titles by various query parameters, which are passed as a `dict`. Similar to `ScopusSearch`, the class consists of only one attribute `SerialSearch.results` returning a list of namedtuples of search results of serial titles.

An example:
```python
>>> from pybliometrics.scopus import SerialSearch
>>> import pandas as pd
>>> query = {'title': 'Neural Computation'}
>>> ser1 = SerialSearch(query, refresh=30)
>>> ser1.results
[Serial(coverageEndYear='2020', coverageStartYear='1995',
 coverimage='https://api.elsevier.com/content/serial/title/issn/0899-7667?view=coverimage',
 dc_publisher='MIT Press Journals', dc_title='Neural Computation',
 homepage='http://www.mitpressjournals.org/openurl?genre=journal&stitle=neco',
 oaAllowsAuthorPaid=None, openArchiveArticle=None, openaccess=None,
 openaccessArticle=None, openaccessStartDate=None, openaccessType=None,
 prism_aggregationType='journal', prism_eIssn='1530-888X', prism_issn='0899-7667',
 prism_url='https://api.elsevier.com/content/serial/title/issn/0899-7667',
 scopus_source='https://www.scopus.com/source/sourceInfo.url?sourceId=24782',
 source_id='24782', subject_area_abbrevs='ARTS;NEUR', subject_area_codes='1201;2805',
 subject_area_names='Cognitive Neuroscience;Arts and Humanities (miscellaneous)',
 SJR_2018='0.885', SNIP_2018='1.142', citeCountSCE_1996='124', citeCountSCE_1997='395',
 citeCountSCE_1998='579', citeCountSCE_1999='934', citeCountSCE_2000='1282',
 citeCountSCE_2001='2024', citeCountSCE_2002='2493', citeCountSCE_2003='3058',
 citeCountSCE_2004='3835', citeCountSCE_2005='4799', citeCountSCE_2006='5403',
 citeCountSCE_2007='6212', citeCountSCE_2008='6504', citeCountSCE_2009='7856',
 citeCountSCE_2010='8137', citeCountSCE_2011='8157', citeCountSCE_2012='9261',
 citeCountSCE_2013='9785', citeCountSCE_2014='10325', citeCountSCE_2015='10755',
 citeCountSCE_2016='11819', citeCountSCE_2017='13259', citeCountSCE_2018='16185',
 citeCountSCE_2019='18629', citeCountSCE_2020='4337', citeScoreCurrentMetric_2018='2.39',
 citeScoreTracker_2019='3.09', publicationCount_1996='93', publicationCount_1997='90',
 publicationCount_1998='105', publicationCount_1999='95', publicationCount_2000='126',
 publicationCount_2001='114', publicationCount_2002='122', publicationCount_2003='112',
 publicationCount_2004='104', publicationCount_2005='107', publicationCount_2006='119',
 publicationCount_2007='125', publicationCount_2008='117', publicationCount_2009='128',
 publicationCount_2010='113', publicationCount_2011='108', publicationCount_2012='118',
 publicationCount_2013='103', publicationCount_2014='100', publicationCount_2015='94',
 publicationCount_2016='97', publicationCount_2017='112', publicationCount_2018='105',
 publicationCount_2019='88', publicationCount_2020='24', revPercent_1996='0',
 revPercent_1997='3.33', revPercent_1998='1.9', revPercent_1999='3.16', revPercent_2000='0.79',
 revPercent_2001='2.63', revPercent_2002='0.82', revPercent_2003='4.46', revPercent_2004='10.58',
 revPercent_2005='2.8', revPercent_2006='0.84', revPercent_2007='1.6', revPercent_2008='2.56',
 revPercent_2009='0', revPercent_2010='0.88', revPercent_2011='0', revPercent_2012='0',
 revPercent_2013='1.94', revPercent_2014='0', revPercent_2015='1.06', revPercent_2016='1.03',
 revPercent_2017='1.79', revPercent_2018='0.95', revPercent_2019='2.27', revPercent_2020='0',
 zeroCitesPercentSCE_1996='0', zeroCitesPercentSCE_1997='1.111111111111111111',
 zeroCitesPercentSCE_1998='0', zeroCitesPercentSCE_1999='3.157894736842105263',
 zeroCitesPercentSCE_2000='0', zeroCitesPercentSCE_2001='3.508771929824561404',
 zeroCitesPercentSCE_2002='0.8196721311475409836',  zeroCitesPercentSCE_2003='0.8928571428571428572',
 zeroCitesPercentSCE_2004='0', zeroCitesPercentSCE_2005='1.869158878504672897',
 zeroCitesPercentSCE_2006='0', zeroCitesPercentSCE_2007='2.4', zeroCitesPercentSCE_2008='0',
 zeroCitesPercentSCE_2009='4.6875', zeroCitesPercentSCE_2010='6.194690265486725664', 
 zeroCitesPercentSCE_2011='1.851851851851851852', zeroCitesPercentSCE_2012='6.779661016949152543', 
 zeroCitesPercentSCE_2013='1.941747572815533981', zeroCitesPercentSCE_2014='4',
 zeroCitesPercentSCE_2015='2.127659574468085106', zeroCitesPercentSCE_2016='11.34020618556701031',
 zeroCitesPercentSCE_2017='19.64285714285714286', zeroCitesPercentSCE_2018='24.76190476190476191', 
 zeroCitesPercentSCE_2019='64.77272727272727273', zeroCitesPercentSCE_2020='95.83333333333333334',
 zeroCitesSCE_1996='0', zeroCitesSCE_1997='1', zeroCitesSCE_1998='0', zeroCitesSCE_1999='3',
 zeroCitesSCE_2000='0', zeroCitesSCE_2001='4', zeroCitesSCE_2002='1', zeroCitesSCE_2003='1',
 zeroCitesSCE_2004='0', zeroCitesSCE_2005='2', zeroCitesSCE_2006='0', zeroCitesSCE_2007='3',
 zeroCitesSCE_2008='0', zeroCitesSCE_2009='6', zeroCitesSCE_2010='7', zeroCitesSCE_2011='2',
 zeroCitesSCE_2012='8', zeroCitesSCE_2013='2', zeroCitesSCE_2014='4', zeroCitesSCE_2015='2',
 zeroCitesSCE_2016='11', zeroCitesSCE_2017='22', zeroCitesSCE_2018='26',
 zeroCitesSCE_2019='57', zeroCitesSCE_2020='23'),
 Serial(coverageEndYear='2020', coverageStartYear='1993', coverimage='https://api.elsevier.com/content/serial/title/issn/0941-0643?view=coverimage', dc_publisher='Springer London', dc_title='Neural Computing and Applications', homepage='https://info.sciverse.com/i.php?issn=0941-0643', oaAllowsAuthorPaid=None, openArchiveArticle=None, openaccess=None, openaccessArticle=None, openaccessStartDate=None, openaccessType=None, prism_aggregationType='journal', prism_eIssn='1433-3058', prism_issn='0941-0643', prism_url='https://api.elsevier.com/content/serial/title/issn/0941-0643', scopus_source='https://www.scopus.com/source/sourceInfo.url?sourceId=24800', source_id='24800', subject_area_abbrevs='COMP', subject_area_codes='1712;1702', subject_area_names='Software;Artificial Intelligence', SJR_2018='0.637', SNIP_2018='1.481', citeCountSCE_1996='40', citeCountSCE_1997='61', citeCountSCE_1998='45', citeCountSCE_1999='55', citeCountSCE_2000='67', citeCountSCE_2001='128', citeCountSCE_2002='139', citeCountSCE_2003='137', citeCountSCE_2004='172', citeCountSCE_2005='225', citeCountSCE_2006='306', citeCountSCE_2007='341', citeCountSCE_2008='470', citeCountSCE_2009='605', citeCountSCE_2010='655', citeCountSCE_2011='885', citeCountSCE_2012='1075', citeCountSCE_2013='1698', citeCountSCE_2014='2471', citeCountSCE_2015='3217', citeCountSCE_2016='4064', citeCountSCE_2017='5424', citeCountSCE_2018='8019', citeCountSCE_2019='12564', citeCountSCE_2020='4814', citeScoreCurrentMetric_2018='4.20', citeScoreTracker_2019='5.58', publicationCount_1996='26', publicationCount_1997='42', publicationCount_1998='32', publicationCount_1999='35', publicationCount_2000='31', publicationCount_2001='27', publicationCount_2002='25', publicationCount_2003='40', publicationCount_2004='41', publicationCount_2005='39', publicationCount_2006='38', publicationCount_2007='55', publicationCount_2008='61', publicationCount_2009='102', publicationCount_2010='113', publicationCount_2011='125', publicationCount_2012='247', publicationCount_2013='507', publicationCount_2014='413', publicationCount_2015='127', publicationCount_2016='208', publicationCount_2017='412', publicationCount_2018='733', publicationCount_2019='1304', publicationCount_2020='288', revPercent_1996='0', revPercent_1997='0', revPercent_1998='0', revPercent_1999='0', revPercent_2000='0', revPercent_2001='3.7', revPercent_2002='4', revPercent_2003='5', revPercent_2004='0', revPercent_2005='2.56', revPercent_2006='2.63', revPercent_2007='0', revPercent_2008='0', revPercent_2009='0', revPercent_2010='0', revPercent_2011='0', revPercent_2012='0.4', revPercent_2013='1.78', revPercent_2014='3.39', revPercent_2015='1.57', revPercent_2016='1.92', revPercent_2017='2.91', revPercent_2018='2.86', revPercent_2019='0.61', revPercent_2020='5.56', zeroCitesPercentSCE_1996='26.92307692307692308', zeroCitesPercentSCE_1997='7.142857142857142857', zeroCitesPercentSCE_1998='9.375', zeroCitesPercentSCE_1999='11.42857142857142857', zeroCitesPercentSCE_2000='3.225806451612903226', zeroCitesPercentSCE_2001='3.703703703703703704', zeroCitesPercentSCE_2002='20', zeroCitesPercentSCE_2003='5', zeroCitesPercentSCE_2004='9.756097560975609756', zeroCitesPercentSCE_2005='7.692307692307692308', zeroCitesPercentSCE_2006='5.263157894736842106', zeroCitesPercentSCE_2007='9.090909090909090909', zeroCitesPercentSCE_2008='8.196721311475409836', zeroCitesPercentSCE_2009='4.901960784313725491', zeroCitesPercentSCE_2010='3.539823008849557522', zeroCitesPercentSCE_2011='8', zeroCitesPercentSCE_2012='6.882591093117408907', zeroCitesPercentSCE_2013='7.692307692307692308', zeroCitesPercentSCE_2014='5.811138014527845037', zeroCitesPercentSCE_2015='7.086614173228346457', zeroCitesPercentSCE_2016='7.211538461538461539', zeroCitesPercentSCE_2017='5.339805825242718447', zeroCitesPercentSCE_2018='15.00682128240109141', zeroCitesPercentSCE_2019='45.70552147239263804', zeroCitesPercentSCE_2020='79.86111111111111111', zeroCitesSCE_1996='7', zeroCitesSCE_1997='3', zeroCitesSCE_1998='3', zeroCitesSCE_1999='4', zeroCitesSCE_2000='1', zeroCitesSCE_2001='1', zeroCitesSCE_2002='5', zeroCitesSCE_2003='2', zeroCitesSCE_2004='4', zeroCitesSCE_2005='3', zeroCitesSCE_2006='2', zeroCitesSCE_2007='5', zeroCitesSCE_2008='5', zeroCitesSCE_2009='5', zeroCitesSCE_2010='4', zeroCitesSCE_2011='10', zeroCitesSCE_2012='17', zeroCitesSCE_2013='39', zeroCitesSCE_2014='24', zeroCitesSCE_2015='9', zeroCitesSCE_2016='15', zeroCitesSCE_2017='22', zeroCitesSCE_2018='110', zeroCitesSCE_2019='596', zeroCitesSCE_2020='230')]
>>> data = pd.DataFrame(ser1.results)
>>> data.shape
(2, 150)
>>> query['field'] = 'title,publisher' # Same query but show only title and publisher
>>> ser2 = SerialSearch(query, refresh=30)
>>> ser2.results
[Serial(dc_publisher='MIT Press Journals', dc_title='Neural Computation', prism_url='https://api.elsevier.com/content/serial/title/issn/0899-7667'),
 Serial(dc_publisher='Springer London', dc_title='Neural Computing and Applications', prism_url='https://api.elsevier.com/content/serial/title/issn/0941-0643')]
>>> query['issn'] = '0899-7667' # Now specify ISSN
>>> query['field'] = 'title,publisher,issn'
>>> ser3 = SerialSearch(query, refresh=30)
>>> ser3.results
[Serial(dc_publisher='MIT Press Journals', dc_title='Neural Computation', prism_issn='0899-7667',
 prism_url='https://api.elsevier.com/content/serial/title/issn/0899-7667')]
```
Some notes:
* Currently, `SerialSearch` allows for all query parameters of [Serial Title API](https://dev.elsevier.com/documentation/SerialTitleAPI.wadl#d1e534) except `content` (search by content type) and `pub` (search by publisher). For reasons yet to be discovered, the search by these parameters does not work. The manual requests to API with `pub` and `content` as request parameters also did not yield expected results.
* Tangential: both Serial Title API methods accept following views STANDARD, ENHANCED, CITESCORE, COVERIMAGE. `SerialTitle` class of pybliometrics allows only BASIC, STANDARD, ENHANCED. Should a correction be made?

SolMe commented:
Update: tests were not functioning properly, as I forgot to account for some recent changes. Fixed

Michael-E-Rose commented on pybliometrics/scopus/__init__.py:
```diff
@@ -9,3 +9,5 @@
 from pybliometrics.scopus.author_search import *
 from pybliometrics.scopus.scopus_search import *
 from pybliometrics.scopus.serial_title import *
+from pybliometrics.scopus.serial_search import *
```
Please order alphabetically

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+        on the length of yearly data Note: can be empty.
+        """
+        out = []
+        search_results = self._json['serial-metadata-response'].get('entry')
+        if search_results:
```
I think if you use `.get('entry', [])`, i.e return an empty list if there are no entries, you can remove one level of hierarchy, which greatly reduces complexity

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+
+
+def _form_fields(results):
+    """Auxiliary function that returns tuple represiting union of keys
```
Docstring contains some typos and should be of the following form:
"""Short one-line description.
empty line
Full description over multiple lines.
"""

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+        if i.get('yearly-data'):
+            yearly_fields = set()
+            for t in i['yearly-data'].get('info', []):
+                period_fields = set(key
+                                    + '_'
+                                    + str(t.get('@year'))
+                                    for key in t.keys()
+                                    if key not in ('@year', '@_fa'))
```
This looks a bit too complicated. Why not `set(f"{k}_{t.get('@year')" for k in t if k not in ('@year', '@_fa'))`? Also, what happens if `t['@year']` returns `None`? Can it actually happen? If no, simply use `t['@year']`.

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+                                    + str(t.get('@year'))
+                                    for key in t.keys()
+                                    if key not in ('@year', '@_fa'))
+                yearly_fields.update(period_fields)
+            time_fields.update(yearly_fields)
```
Why maintaining two sets where one goes into the other one anyways?

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+        # Link names
+        links = [j['@ref'] for j in i.get('link', []) if j.get('@ref')]
+        base_fields.update(set(links))
+        if i.get('citeScoreYearInfoList'):
```
Rather than testing the return value, which are two operations, simply check `'citeScoreYearInfoList' in i`. That's just one operation. Same for line 202.

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+
+
+def _clean_field_name(fieldname):
+    """Cleans string for use as namedtuple field"""
```
No third person language in function docstrings please.

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -0,0 +1,303 @@
+            stat_val = t.get('$')
+            out.append([stat_name, stat_val])
+    return out or None
+            
```
Seems the last line is not empty

Michael-E-Rose commented on pybliometrics/scopus/superclasses/search.py:
```diff
@@ -60,9 +60,18 @@ def __init__(self, query, api, refresh, view='STANDARD', count=200,
+            name = query
+            params['query'] = query
+        else:
+            name = ''
+            params.update(query)
+            for q in query:
+                name += q + ' ' + query[q] + '&'
+            name = name[:-1]
```
Replace by `name = "&".join(["=".join(t) for t in zip(query.keys(), query.values())])`. Faster and easier to read

Michael-E-Rose commented on pybliometrics/scopus/utils/constants.py:
```diff
@@ -10,7 +10,8 @@
     'ContentAffiliationRetrieval': expanduser('~/.scopus/affiliation_retrieval'),
     'ScopusSearch': expanduser('~/.scopus/scopus_search'),
     'SerialTitle': expanduser('~/.scopus/serial_title'),
-    'PlumXMetrics': expanduser('~/.scopus/plumx')
+    'PlumXMetrics': expanduser('~/.scopus/plumx'),
+    'SerialSearch': expanduser('~/.scopus/serial_search')
```
Alphabetic order please

SolMe commented:
Following changes were introduced:
* Removed namedtuples. Instead `SerialSearch.results` outputs list of OrderedDicts. Accordingly, functions `_clean_field_name` and `_form_fields` were removed.
* Corrected docstrings: one-line description followed by blank line and full description. No third person language.
* Tests were changed as well with removal of namedtuples from `SerialSearch.results`.
* Minor changes: `.get('entry', [])` in `serial_search.py`, entries in alphabetical order, change of turning query dict to string as requested, etc.

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -1,49 +1,44 @@
+            # OrderedDict to populate with individual serial data
+            obs = OrderedDict()
+            for key in i:
+                if key != '@_fa' and i[key]:
```
One can remove one level by inverting this statement followed by `continue`, i.e.
```
if key == '@_fa':
    continue
<normal code>
```
What's the purpose of testing `i[key]`? It's the element we're iterating over, isn't?

SolMe replied:
Some keys can have a `None` value, hence testing.

SolMe replied:
Here is an example of a result dictionary with keys with `None` values:
```python
{'@_fa': 'true',
 'dc:title': '10th International Symposium on Integrated Circuits, Devices and Systems, ISIC-2004: Integrated Systems on Silicon - Proceedings',
 'dc:publisher': None,
 'coverageStartYear': '2004',
 'coverageEndYear': '2004',
 'prism:aggregationType': 'conferenceproceeding',
 'source-id': '78598',
 'openaccess': None,
 'openaccessArticle': None,
 'openArchiveArticle': None,
 'openaccessType': None,
 'openaccessStartDate': None,
 'oaAllowsAuthorPaid': None,
 'subject-area': [{'@_fa': 'true',
   '@code': '2200',
   '@abbrev': 'ENGI',
   '$': 'Engineering (all)'}],
 'SNIPList': {'SNIP': [{'@_fa': 'true', '@year': '2007', '$': '0.032'}]},
 'SJRList': {'SJR': [{'@_fa': 'true', '@year': '2007', '$': '0.102'}]},
 'link': [{'@_fa': 'true',
   '@ref': 'scopus-source',
   '@href': 'https://www.scopus.com/source/sourceInfo.url?sourceId=78598'}],
 'yearly-data': None}
```
The most problematic part are keys with dictionary values (e.g. `yearly-data`). Normally, the `yearly-data` key from the above example would look like this:
```python
{...,
 'yearly-data': {'info': [{'@_fa': 'true',
    '@year': '1996',
    'publicationCount': '0',
    'citeCountSCE': '0',
    'zeroCitesSCE': '0',
    'zeroCitesPercentSCE': None,
    'revPercent': '0'},
   {'@_fa': 'true',
    '@year': '1997',
    'publicationCount': '0',
    'citeCountSCE': '0',
    'zeroCitesSCE': '0',
    'zeroCitesPercentSCE': None,
    'revPercent': '0'}]}}
```
Testing for existence of keys is not enough. Since keys with container values are passed through auxiliary functions, in this case we would be searching for `info` key using `result['yearly-data'].get('info')` which will return `AttributeError` on `None`. The solution is to also test value of each key.
Alternatively, we could test value of each key with container value but that would mean adding additional `if` clauses.

Michael-E-Rose replied:
I see. Then let's do like this:
``````

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -1,49 +1,44 @@
+        for i in search_results:
+            # OrderedDict to populate with individual serial data
+            obs = OrderedDict()
+            for key in i:
```
If `i` is a dict, we should do three things:
1. in line 14, the default value should be `{}`, not `[]`
2. in line 18, we should iterate over `i` using both key _and_ value, so to not use `i[key]` in the loop
3. since `i` is rather reserved for indices, we should rename it to make clear it's an element in a list, e.g. `item` (I think that's what other classes are using)

SolMe replied:
1. I do not see why default value should be `{}` and not `[]`. The `search_results` is list of dictionaries.

Michael-E-Rose replied:
Oh yes, you're absolutely right!

Michael-E-Rose commented on pybliometrics/scopus/superclasses/search.py:
```diff
@@ -67,9 +67,7 @@ def __init__(self, query, api, refresh, view='STANDARD', count=200,
         else:
             name = ''
```
Due to the new line 70 this is not now superfluous

SolMe replied:
Did you mean "this is now superfluous"?

Michael-E-Rose replied:
Yes. It's not necessary to instantiate a str variable any longer. You used this variable to append to, but now the variable is created in one go

SolMe commented:
All requested changes were incorporated. I have also deleted the `verbose` keyword argument. It does not apply, since technically `SerialSearch` is a retrieval method.

Michael-E-Rose commented on pybliometrics/scopus/serial_search.py:
```diff
@@ -12,37 +12,46 @@ def results(self):
+                elif key == 'yearly-data':
+                    if not value:
+                        continue
+                    time_data = _retrieve_yearly_data(value.get('info'))
+                    for j in time_data:
```
I think this might break and result in an attribute error if the returned value is `None` (which is the default from `.get()`). Do we have reason to believe that the "info" key might be missing even though "yearly-data" is present? If yes, `.get()`'s default should be an empty list. If not, let's simply use `value['info']`.

Michael-E-Rose commented:
Outstanding work! Glad to have you as a contributor!
